### PR TITLE
[WIP] Port compute_diag third_shoc_moment

### PR DESCRIPTION
**Pull request by tcclevenger**
Port `compute_diag_third_shoc_moment()`. This function depends on `fterms_input_for_diag_third_shoc_moment`, `f0_to_f5_diag_third_shoc_moment`, `omega_terms_diag_third_shoc_moment`, `x_y_terms_diag_third_shoc_moment`, `aa_terms_diag_third_shoc_moment`, `w3_diag_third_shoc_moment`, but all those functions are only called in `compute_diag_third_shoc_moment`, so I chose not to implement each of those functions in separate files. Let me know if people prefer separate implementations. 

If not, should I delete all the tests for the non-implemented function? Or at least delete the bfb portion of the tests? E.g. `shoc_fterm_input_third_moms_tests.cpp`.

This PR depends on https://github.com/E3SM-Project/scream/pull/599, but I imagine that PR will be merged much sooner. Also, I've labeled [WIP] till the testers are back up.

### CODE REVIEW

File: components/scream/src/physics/shoc/shoc_functions_f90.cpp
Comment (tcclevenger) [inline components/scream/src/physics/shoc/shoc_functions_f90.cpp:781]: Should these inputs be removed further up the call stack?

File: components/scream/src/physics/shoc/shoc_functions_f90.cpp
Comment (tcclevenger) [inline components/scream/src/physics/shoc/shoc_functions_f90.cpp:780]: Should these inputs be removed further up the call stack?

Comment (ambrad) [reply components/scream/src/physics/shoc/shoc_functions_f90.cpp:780]: Not used in either impl, right? Then just rm them from F90. That's BFB, so can do that in this PR.

File: components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp
Comment (ambrad) [inline components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp:71]: Is masking `range_pack1 < 1` right even when the shift is +1? Seems to me the 0 -> 1 index shift will be wrong.

File: components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp
Comment (ambrad) [inline components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp:44]: No need for these decls. Just use `const auto f1 = ` for example.

File: components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp
Comment (tcclevenger) [reply components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp:71]: The loop is `k = 2,...,nlev` in shoc.F90, (switching to c++ 0 based units) so we never set actually use the 0->1 shift.

File: components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp
Comment (jgfouca) [inline components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp:40]: Always prefer integer literals if you're dealing with a round number so the compiler can promote it to the appropriate type. I see a lot of `X.0` literals in the code below. I think all of them can be `X`

**Review by jgfouca — CHANGES_REQUESTED** [review]
Minor changes needed for literals, otherwise excellent job! This conversion looks pretty challenging.

File: components/cam/src/physics/cam/shoc.F90
Comment (jeff-cohere) [inline components/cam/src/physics/cam/shoc.F90:1720]: This change isn't BFB, is it?

Comment (jeff-cohere) [reply components/cam/src/physics/cam/shoc.F90:1720]: In the case that you're changing the Fortran to be BFB-equal to C++, this creates a non-BFB PR that should arguably be separate from the C++ conversion work.

Comment (jeff-cohere) [reply components/cam/src/physics/cam/shoc.F90:1720]: Or maybe I'm misunderstanding, and all these changes went through in a prior PR? If so, please ignore this comment.

Comment (tcclevenger) [reply components/cam/src/physics/cam/shoc.F90:1720]: These changes are included in https://github.com/E3SM-Project/scream/pull/599. Once it is merged and I rebase to master these will not be here. Should I change to [WIP] until that PR is merged?

File: components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp
Comment (tcclevenger) [reply components/scream/src/physics/shoc/shoc_compute_diag_third_shoc_moment_impl.hpp]: Ok, good to know! I've made the change.

File: components/cam/src/physics/cam/shoc.F90
Comment (jeff-cohere) [reply components/cam/src/physics/cam/shoc.F90:1720]: No, that's fine. I'm just juggling enough things to be distracted, is all. :-) Sorry for the noise.

Comment (tcclevenger) [reply components/cam/src/physics/cam/shoc.F90:1720]: No worries at all! I have so many interdependent PRs going on.. ha

**Review by jeff-cohere — APPROVED** [review]
Nice work! I think collecting all of the Fortran subroutines here is definitely justified in the amount of boilerplate it saves.

**Comment by tcclevenger** [comment]
Working off a forked branch currently, which doesn't allow the tester to run without approval. Switching to branch off of the scream repo so I can debug without everyone getting emails. I'll request reviews once I've sorted out the issues.

**Pull request closed without merging**
